src/main/groovy/lang/ObjectRange.java

```diff
@@ -493,40 +495,52 @@ public void remove() {
 
     /**
      * Increments by step size
+     * <p/>
+     * It uses arithmetic plus for {@link Number} instances and the repeated `next` method for other {@link Comparable}s.
      *
      * @param value the value to increment
      * @param step  the value to increment by
      * @return the incremented value or null, if there isn't any
      */
     @SuppressWarnings("unchecked")
     private T increment(T value, int step) {
-        for (int i = 0; i < step; i++) {
-            final T next = (T) InvokerHelper.invokeMethod(value, "next", null);
-            if (!compareGreaterThan(next, value)) /* e.g. `next` of the last element */ {
-                return null;
+        if (value instanceof Number) {
+            return (T) plus((Number) value, step);
+        } else {
+            for (int i = 0; i < step; i++) {
+                final T next = (T) InvokerHelper.invokeMethod(value, "next", null);
+                if (!compareGreaterThan(next, value)) /* e.g. `next` of the last element */ {
+                    return null;
+                }
+                value = next;
             }
-            value = next;
+            return value;
         }
-        return value;
     }
 
     /**
      * Decrements by step size
+     * <p/>
+     * It uses arithmetic minus for {@link Number} instances and the repeated `previous` method for other {@link Comparable}s.
      *
      * @param value the value to decrement
      * @param step  the value to decrement by
      * @return the decremented value or null, if there isn't any
      */
     @SuppressWarnings("unchecked")
     private T decrement(T value, int step) {
-        for (int i = 0; i < step; i++) {
-            final T previous = (T) InvokerHelper.invokeMethod(value, "previous", null);
-            if (!compareLessThan(previous, value)) /* e.g. `previous` of the first element */ {
-                return null;
+        if (value instanceof Number) {
+            return (T) minus((Number) value, step);
+        } else {
+            for (int i = 0; i < step; i++) {
+                final T previous = (T) InvokerHelper.invokeMethod(value, "previous", null);
+                if (!compareLessThan(previous, value)) /* e.g. `previous` of the first element */ {
+                    return null;
+                }
+                value = previous;
             }
```

Comment:
So the main point of consideration now is do we want this breaking change? If so, how we might document it? Is it possible (and would it be worthwhile) to provide a compatibility story or some kind of support for people wanting the old behavior?